airbyte-integrations/connectors/source-pipedrive/source_pipedrive/streams.py

```diff
@@ -0,0 +1,208 @@
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, Mapping, MutableMapping, Optional
+
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+
+PIPEDRIVE_URL_BASE = "https://api.pipedrive.com/v1/"
+
+
+class PipedriveStream(HttpStream, ABC):
+    url_base = PIPEDRIVE_URL_BASE
+    primary_key = "id"
+    data_field = "data"
+
+    def __init__(self, api_token: str, **kwargs):
+        super().__init__(**kwargs)
+        self._api_token = api_token
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        class_name = self.__class__.__name__
+        return f"{class_name[0].lower()}{class_name[1:]}"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return {"api_token": self._api_token}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        records = response.json().get(self.data_field) or []
+        yield from records
+
+
+class PaginatedPipedriveStream(PipedriveStream):
+    limit = 50
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """
+        :param response: the most recent response from the API
+        :return If there is another page in the result, a mapping (e.g: dict) containing information needed to query the next page in the response.
+                If there are no more pages in the result, return None.
+        """
+        pagination_data = response.json().get("additional_data", {}).get("pagination", {})
+        if pagination_data.get("more_items_in_collection"):
+            start = pagination_data.get("start", 0) + self.limit
```

Comment:
Are you sure that `start` will return to us in response, otherwise we will have an infinite loop?

Reply:
Well, yes, because we always have `start` and `limit` keys, if `more_items_in_collection=True`. But I'll add and extra conditions to double check.